addons/account_audit_trail/models/account_move.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import logging
+
+
+from odoo import api, models, _
+from odoo.exceptions import UserError
+
+_logger = logging.getLogger(__name__)
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_account_audit_trail_except_once_post(self):
+        if not self._context.get('force_delete') and any(
+            move.posted_before and move.company_id.check_account_audit_trail
+            for move in self
+        ):
+            raise UserError(_("To keep the audit trail, you can not delete journal entries once they have been posted.\nInstead, you can cancel the journal entry."))
+
+    def unlink(self):
+        # Add logger here because in api ondelete account.move.line is deleted and we can't get total amount
+        logger_msg = False
+        if any(m.posted_before and m.company_id.check_account_audit_trail for m in self):
+            if self._context.get('force_delete'):
+                moves_details = []
+                for move in self:
+                    entry_details = "{move_name} ({move_id}) amount {amount_total} {currency} and partner {partner_name}".format(
+                        move_name=move.name,
+                        move_id=move.id,
+                        amount_total=move.amount_total,
+                        currency=move.currency_id.name,
+                        partner_name=move.partner_id.display_name,
+                    )
+                    account_balances_per_account = {}
+                    for line in move.line_ids:
+                        account_id = line.account_id.id
+                        account_balances_per_account[account_id] = account_balances_per_account.get(account_id, 0.0) + line.balance
+                    account_details = "\n".join("- {account_name} ({account_id}) with balance {balance} {currency}".format(
+                            account_name=self.env['account.account'].browse(account_id).name,
```

Comment:
Can't use ```line.account_id.name```, because it will return only last line's ```account_id.name```

Reply:
Ok, then something like this
```python
                    account_balances_per_account = defaultdict(float)
                    for line in move.line_ids:
                        account = line.account_id
                        account_balances_per_account[account] += line.balance
                    account_details = "\n".join(
                        "- {account_name} ({account_id}) with balance {balance} {currency}".format(
                            account_name=account.name,
                            account_id=account.id,
                            balance=balance,
                            currency=line.currency_id.name,
                        )
                        for account, balance in account_balances_per_account.items()
                    )
```

* use defaultdict to avoid reading the previous value and longer lines
* use the record instead of the id to avoid rebrowsing (otherwise you will lose the prefetch ids)
* have consistent parenthesis opening/closing (they must be on the same indentation level for better readability)